micropython/drivers/imu/lsm9ds1/lsm9ds1.py

```diff
@@ -58,129 +58,149 @@
-        # ctrl9 - FIFO enabled
-        mv[0] = mv[1] = 0x38
-        mv[2] = ((sample_rate & 7) << 5) | ((_SCALE_ACCEL[scale_accel][1] & 0x3) << 3)
-        mv[3] = 0x00
-        mv[4] = 0x4
-        mv[5] = 0x2
-        i2c.writeto_mem(addr, _CTRL_REG4_G, mv[:6])
+        mv[2] = (ODR_IMU.index(accel_odr) << 5) | ((ACCEL_SCALE.index(accel_scale)) << 3)
+        mv[3] = 0x00  # ctrl7 - leave at default values
+        mv[4] = 0x4  # ctrl8 - leave at default values
+        mv[5] = 0x2  # ctrl9 - FIFO enabled
+        self.bus.writeto_mem(self.address_imu, _CTRL_REG4_G, mv[:6])
 
         # fifo: use continous mode (overwrite old data if overflow)
-        i2c.writeto_mem(addr, _FIFO_CTRL_REG, b"\x00")
-        i2c.writeto_mem(addr, _FIFO_CTRL_REG, b"\xc0")
+        self.bus.writeto_mem(self.address_imu, _FIFO_CTRL_REG, b"\x00")
+        self.bus.writeto_mem(self.address_imu, _FIFO_CTRL_REG, b"\xc0")
 
-        self.scale_gyro = 32768 / _SCALE_GYRO[scale_gyro][0]
-        self.scale_accel = 32768 / _SCALE_ACCEL[scale_accel][0]
-
-    def init_magnetometer(self, sample_rate=7, scale_magnet=0):
-        """
-        sample rates = 0-7 (0.625, 1.25, 2.5, 5, 10, 20, 40, 80Hz)
-        scaling = 0-3 (+/-4, +/-8, +/-12, +/-16 Gauss)
-        """
-        assert sample_rate < 8, "invalid sample rate: %d (0-7)" % sample_rate
-        assert scale_magnet < 4, "invalid scaling: %d (0-3)" % scale_magnet
-        i2c = self.i2c
-        addr = self.address_magnet
-        mv = memoryview(self.scratch)
-        mv[0] = 0x40 | (sample_rate << 2)  # ctrl1: high performance mode
-        mv[1] = scale_magnet << 5  # ctrl2: scale, normal mode, no reset
+        # Configure Magnetometer
+        mv[0] = 0x40 | (magnet_odr << 2)  # ctrl1: high performance mode
+        mv[1] = MAGNET_SCALE.index(magnet_scale) << 5  # ctrl2: scale, normal mode, no reset
         mv[2] = 0x00  # ctrl3: continous conversion, no low power, I2C
         mv[3] = 0x08  # ctrl4: high performance z-axis
         mv[4] = 0x00  # ctr5: no fast read, no block update
-        i2c.writeto_mem(addr, _CTRL_REG1_M, mv[:5])
-        self.scale_factor_magnet = 32768 / ((scale_magnet + 1) * 4)
+        self.bus.writeto_mem(self.address_magnet, _CTRL_REG1_M, mv[:5])
+
+        self.gyro_scale = 32768 / gyro_scale
+        self.accel_scale = 32768 / accel_scale
+        self.scale_factor_magnet = 32768 / ((MAGNET_SCALE.index(magnet_scale) + 1) * 4)
+
+        # Allocate scratch buffer for efficient conversions and memread op's
+        self.scratch_int = array.array("h", [0, 0, 0])
 
     def calibrate_magnet(self, offset):
         """
         offset is a magnet vecor that will be substracted by the magnetometer
         for each measurement. It is written to the magnetometer's offset register
         """
         offset = [int(i * self.scale_factor_magnet) for i in offset]
-        mv = memoryview(self.scratch)
+        mv = memoryview(array.array("B", [0, 0, 0, 0, 0, 0]))
```

Comment:
This doesn't need a memoryview as it always either subscripts or slices the full length. Also can be `bytearray(6)` instead of array.

On second thought, could this function just do

```
self.bus.writeto_mem(self.address_magnet, _OFFSET_REG_X_M, struct.pack("<H", offset))
```

Reply:
> On second thought, could this function just do

Probably yes (you mean `<HHH` ?), but would have to import struct and this function is not used from within the driver.

Reply:
Yep `<HHH` sorry.
There's very little cost to `import struct`, and you'll end up with shorter code, therefore less bytecode (i.e. less RAM overall if running from .py or .mpy).
Plus you can defer the `import struct` to inside the function that uses it.